rfcs/20220307-monitoring_across_the_stack.md

```diff
@@ -0,0 +1,355 @@
+
+Prometheus is a time series database with an advanced query language and the ability to configure aggregate data transformations (for example, based on a given metric you can also tell Prometheus to calculate a running mean of the metric from over the last 5 minutes, or any other span of time). It includes anomaly detection as well as integration with many tools you can use for alerting like Slack, PagerDuty, or email. There are [official and unofficial client libraries](https://prometheus.io/docs/instrumenting/clientlibs/) for all the langauges we currently use or could conceivably use in the future including Python, Node.js, Go, PHP, and JVM languages.
+
+Grafana is a data visualization tool that supports building complex dashboards. It has the ability to integrate with nearly any source of data you could imagine, including our existing catalog and API databases. In the future we could use it to monitor cataloging progress over time for example. It's a tool good for many things in addition to visualizing metrics.
+
+The Prometheus and Grafana services encompass the following key features that are important for maintaining the health and stability of our services:
+
+* Metrics gathering
+* Dashboard creation for making it quick and easy to understand the current state of our service health
+* Anomaly detection and alarming so that if things get weird, we'll know about it as soon as possible
+
+Additionally, both are free software. Prometheus is Apache 2 licensed and Grafana is AGPL-3.0. Both are actively developed in the open with heavy community involvement. They are also both widely deployed with lots of blog posts discussing configuration, use cases, and troubleshooting.
+
+## Self-hosted vs. hosted
+
+We will self-host and use a new Postgres RDS instance for Grafana.
+
+To simplify the hosting situation, let's try to deploy this using ECS from the beginning.
+
+## What will we monitor
+
+There are three key aspects of the application services I think we should monitor, along with a slew of system level metrics that we should monitor as well.
+
+### System level metrics
+
+* CPU
+* RAM
+* Storage
+* Processes/threads
+
+These would be gathered from Airflow, API, Thumbnail proxy, Postgres and Nuxt. These metrics need to be retrieved from AWS. I _think_ that [this page describes how to do that](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-PrometheusEC2.html) but AWS's language around this stuff is so cryptic to me I can't say with confidence it's the right documentation page for it.
+
+It would also be nice if we could integrate Cloudflare data into Grafana so that our visualizations are all in once place. Cache hits and other relevant information from Cloudflare would be excellent to monitor within our greater monitoring stack.
```

Comment:
```suggestion
It would also be nice if we could integrate Cloudflare data into Grafana so that our visualizations are all in one place. Cache hits and other relevant information from Cloudflare would be excellent to monitor within our greater monitoring stack.
```